feat(cms): add post answer lists, likes, and shared access utils

nickhsine opened the pull request:
### Summary
  - Add PostChoiceAnswer and PostEssayAnswer lists and wire them to question lists.
  - Introduce PostEssayAnswerLike for “likes” tracking and expose likesCount.
  - Centralize ACL with member-owned access helpers.
  - Update schemas and apply explicit Keystone TypeInfo typings.
### Changes
  - Lists
      - PostChoiceAnswer: relate to PostChoiceQuestion + Member; auto-score via question options.
      - PostEssayAnswer: relate to PostEssayQuestion + Member; add virtual likesCount.
      - PostEssayAnswerLike: join list (answer, member) with unique compositeKey; timestamps.
  - Access control
      - Add utils: memberOwnedOperationAccess, makeMemberOwnedFilter('member' | 'self').
      - Refactor Member/PostEssayAnswer/PostChoiceAnswer to use shared ACL.
  - Typing
      - Apply explicit list typings via TypeInfo across lists.
  - Schema
      - Update schema.graphql/schema.prisma and add migration for post question lists.
### Notes
  - Migration: run prisma migrate to create new tables and relations.
  - Env vars: OpenAI org/project keys already integrated earlier in branch; ensure CMS envs are set.
### Manual QA
  - CMS: create PostChoiceQuestion/PostEssayQuestion; create answers; verify auto-scoring (choice) and likesCount (essay).
  - ACL: as Member, only see/edit own answers; as Admin, full access.
  - Likes: create PostEssayAnswerLike for a member/answer pair; duplicates should be prevented by compositeKey.

### 實作細節
在處理 answer 的 like 數量時，有兩種可行的方案，一是在 PostEssayAnswer list 中，直接加上 `likes` 欄位，並且關聯 Member list，例如：
```
// lists/post-essay-answer.tsx

likes: relationship({
  label: '按讚數量',
  ref: 'Member',
  many: true,
})
```

上述的寫法 Keystone 會產生 implicit many-to-many 的 relationship table，但這個 table 由 Keystone 掌控，我們無法在該 table 加入其他資訊，例如 `created_at`；所以我們僅能紀錄哪些 member 按讚哪些 answer，但是不知道什麼時候按讚的。

為了讓我們掌握度高一點，我選擇了現在 PR 裡的方案，建立一個 PostEssayAnswerLike list，裡面的每一個 record 紀錄「某一個 member 和某一個 anwser 的 like 關係」，因為 PostEssayAnswerLike list 是我們自己定義的，所以可以在其增加其他的 metadata，像是 `created_at`；但是 tradeoff 就是要自己管理 PostEssayAnswerLIke list。

nickhsine committed:
feat(cms): add PostChoiceAnswer list linked to PostChoiceQuestion

- Create lists/post-choice-answer.ts with fields:
    - question (rel: PostChoiceQuestion), member (rel: Member)
    - choiceIndex, correct, compositeKey
    - createdAt, updatedAt
- Access control:
    - Ops: query/create/update/delete for Admin/Member
    - Filter: members limited to query/update/delete own records
- Auto-scoring:
    - Compute correct by comparing choiceIndex with related PostChoiceQuestion.options
- Uniqueness:
    - Use compositeKey = {questionId}:{memberId} to avoid duplicates

nickhsine committed:
feat(cms): add PostEssayAnswer list linked to PostEssayQuestion

- Create lists/post-essay-answer.ts with fields:
  - question (rel: PostEssayQuestion), member (rel: Member)
  - content, compositeKey (unique), createdAt, updatedAt
- Access control:
  - Ops: create/update/delete for Admin/Member, query for
    Admin/Member/FrontendHeadlessAccount
  - Filter: members limited to update/delete own records
- Uniqueness:
  -Use compositeKey = {questionId}:{memberId} to avoid duplicates

nickhsine committed:
refactor(cms): extract shared member-owned access utilities

- Add lists/utils/member-owned-access.ts exporting:
  - memberOwnedOperationAccess for Admin/Member operations
  - makeMemberOwnedFilter(relationKey) to filter by current member
- Replace duplicated access code in:
  - lists/post-choice-answer.ts
  - lists/post-essay-answer.ts

nickhsine committed:
refactor(cms): use shared member-owned access in Member list

- Replace inline access with shared utils:
  - operation: memberOwnedOperationAccess
  - filter: makeMemberOwnedFilter('self')
- Simplify field-level ACL:
  - Remove explicit read rules on twreporter_user_id, role, twoFactorAuth
  - Preserve create/update: false for these fields
- Tweak UI modes for clarity:
  - twreporter_user_id: create hidden, item read, list hidden
  - role: create hidden, item read, list hidden

nickhsine committed:
feat(cms): add PostEssayAnswerLike list for likes tracking

- Create lists/post-essay-answer-like.ts with fields:
  - answer (rel: PostEssayAnswer), member (rel: Member)
  - compositeKey unique, createdAt, updatedAt
- Access control:
  - Operations: memberOwnedOperationAccess
  - Filters: makeMemberOwnedFilter('member')
- Hooks:
  - Set compositeKey = ${answerId}:${memberId} in resolveInput
- UI:
  - Hide create fields; show minimal columns in list

nickhsine committed:
feat(cms): add likesCount to PostEssayAnswer

-  Add likesCount virtual field resolving count via
   context.db.PostEssayAnswerLike.count
- UI: read-only for item/list views

nickhsine committed:
chore(cms): update schema.(grapql,prisma)

nickhsine committed:
refactor(cms): apply explicit typings to lists

- Add/align generated Keystone types across lists:
  - post-choice-answer.ts
  - post-choice-question.ts
  - post-essay-answer-like.ts
  - post-essay-answer.ts
  - post-essay-question.ts
  - member.ts

nickhsine commented on packages/cms/lists/utils/member-owned-access.ts:
```diff
@@ -0,0 +1,27 @@
+
+// Common operation access for member-owned resources
+export const memberOwnedOperationAccess = allowRoles([
+  RoleEnum.Admin,
```
update PR description.

nickhsine committed:
refactor(cms): update PostEssayAnswer `compositeKey.db.isNullable: true`

nickhsine committed:
chore(cms): update schema.(graphql,prisma)

nickhsine committed:
refactor(cms): context.sudo() to bypass PostEssayQuestion list ACL

nickhsine committed:
refactor(cms): deny write access and omit gql for create/update compositeKey

nickhsine commented on packages/cms/lists/post-essay-answer.ts:
```diff
@@ -0,0 +1,133 @@
+        async resolve(item, args, context) {
+          const answerId = item.id
+
+          // Intentionally bypasses PostEssayAnswerLike list ACL via Prisma.
+          // Make sure this resolver already enforced authorization.
+          const count = await context.sudo().db.PostEssayAnswerLike.count({
+            where: { answer: { id: { equals: answerId.toString() } } },
+          })
+
+          return count ?? 0
```
Note:
經測試，copilot 建議的這個做法和 https://github.com/kids-reporter/kids-reporter-monorepo/pull/788#discussion_r2432750803 做法，都沒有辦法在同一個 context 中拿到多個 answerIds。
如果要解決 N+1 pattern 的問題，可能還是會需要寫 custom GQL query 來處理。

關於這個優化，我打算另開一個 PR 處理。

nickhsine commented on packages/cms/lists/post-essay-answer.ts:
```diff
@@ -0,0 +1,133 @@
+        hideCreate: true,
+      },
+    }),
+    content: text({
+      label: '內容',
+      validation: { isRequired: true },
+    }),
```
已經將更新過的 schema.prisma 檔案上傳。

nickhsine committed:
chore(cms): update schema.graphql

nickhsine committed:
refactor(cms): allow create and edit answers and likes for UI testing

Since frontend integration is not ready, we need admin UI to create and
edit answers (PostChoiceAnswer and PostEssayAnswer) and likes
(PostEssayAnswerLike) for testing.
Those admin UI configurations could be enabled if frontend integration
is ready.

nickhsine commented on packages/cms/lists/post-choice-answer.ts:
```diff
@@ -0,0 +1,135 @@
+        const correctIndex = q?.options?.findIndex(
+          (o: { isCorrectAnswer: boolean }) => o.isCorrectAnswer === true
+        )
+        resolvedData.correct = choiceIndex === correctIndex
```
done [d95eb6c](https://github.com/kids-reporter/kids-reporter-monorepo/pull/788/commits/d95eb6c06f37fae4100213b7eff6e2f401cad9b7)

nickhsine commented on packages/cms/lists/post-choice-answer.ts:
```diff
@@ -0,0 +1,135 @@
+    }),
+    correct: checkbox({
+      label: '是否答對',
+      defaultValue: false,
```
done [9caa358](https://github.com/kids-reporter/kids-reporter-monorepo/pull/788/commits/9caa3587c664ee5a76cb034d57a4b71f51cf4382)

nickhsine commented on packages/cms/schema.prisma:
```diff
@@ -334,38 +334,88 @@ model OnlineUser {
+  questionId                      Int?                  @map("question")
+  member                          Member?               @relation("PostEssayAnswer_member", fields: [memberId], references: [id])
+  memberId                        String?               @map("member")
+  compositeKey                    String                @unique @default("")
```
done https://github.com/kids-reporter/kids-reporter-monorepo/pull/788/commits/57435fdadba6d7638cbe347c292303408a841470

nickhsine commented on packages/cms/lists/post-essay-answer-like.ts:
```diff
@@ -0,0 +1,83 @@
+      isIndexed: 'unique',
+      db: {
+        isNullable: true,
+      },
```
done https://github.com/kids-reporter/kids-reporter-monorepo/pull/788/commits/12e637b987fff35d4b62855a88d93d95fc4d88f2

nickhsine commented on packages/cms/lists/post-essay-answer-like.ts:
```diff
@@ -0,0 +1,83 @@
+      many: false,
+      ui: { hideCreate: true },
+    }),
+    compositeKey: text({
+      label: '唯一鍵',
```
done https://github.com/kids-reporter/kids-reporter-monorepo/pull/788/commits/12e637b987fff35d4b62855a88d93d95fc4d88f2

nickhsine commented:
conflicts 的部分等 PR review 完再解，因為怕需要 force push commits 導致 commit hash 跑掉。

nickhsine commented on packages/cms/lists/post-essay-answer.ts:
```diff
@@ -0,0 +1,133 @@
+    likesCount: virtual({
+      field: graphql.field({
+        type: graphql.Int,
+        async resolve(item, args, context) {
+          const answerId = item.id
+
+          // Intentionally bypasses PostEssayAnswerLike list ACL via Prisma.
+          // Make sure this resolver already enforced authorization.
+          const count = await context.sudo().db.PostEssayAnswerLike.count({
+            where: { answer: { id: { equals: answerId.toString() } } },
+          })
+
+          return count ?? 0
```
見 https://github.com/kids-reporter/kids-reporter-monorepo/pull/788#discussion_r2436495200

nickhsine commented on packages/cms/lists/member.ts:
```diff
@@ -1,28 +1,13 @@
+const operationAccessControl = memberOwnedOperationAccess
+const filterAccessControl = makeMemberOwnedFilter('self')
 
 const listConfigurations: ListConfig<any> = list({
```
`ListConfig<any>` -> `ListConfig<ListType<'PostChoiceAnswer'>>`

ericsen-tsai commented on packages/cms/lists/utils/member-owned-access.ts:
```diff
@@ -0,0 +1,27 @@
+// otherwise → filter by relationKey.id equals member id
+export const makeMemberOwnedFilter =
+  (relationKey: 'self' | string) =>
+  ({ session }: { session?: any }) => {
```
nitpick:
define the minimal type for the session instead of `any`

```typescript
type MemberSession = {
  data?: {
    role?: (typeof RoleEnum)[keyof typeof RoleEnum]
    member?: {
      id: string
    }
    memberId?: string
  }
}
```

nickhsine replied:
Addressed in [0a07a3a](https://github.com/kids-reporter/kids-reporter-monorepo/pull/788/commits/0a07a3a5463f9508907c6e92a7ac66a844ecc523)

ericsen-tsai commented on packages/cms/lists/utils/member-owned-access.ts:
```diff
@@ -0,0 +1,27 @@
+// relationKey === 'self' → filter by own id
+// otherwise → filter by relationKey.id equals member id
+export const makeMemberOwnedFilter =
+  (relationKey: 'self' | string) =>
```
nitpick:
use `'self' | (string | {})` to make typescript enable autocomplete for "self"

ref: https://stackoverflow.com/questions/74467392/autocomplete-in-typescript-of-literal-type-and-string

ericsen-tsai commented on packages/cms/lists/utils/member-owned-access.ts:
```diff
@@ -0,0 +1,27 @@
+  (relationKey: 'self' | string) =>
+  ({ session }: { session?: any }) => {
+    const role = session?.data?.role
+    if (role === RoleEnum.Admin) {
```
question:
Do we have to bypass the access for `RoleEnum.Owner` now?

nickhsine replied:
實務上，因為 assign Owner 的 user 目前只有一個人，但是在寫 access control list 時，常常會需要多寫 Owner 的權限設定。我有點想偷懶，所以想說之後就設定到 Admin 權限就好。你覺得會需要保留 Owner 的權限設定嗎？

ericsen-tsai replied:
OK, I think we can only check Admin now!

ericsen-tsai commented on packages/cms/lists/post-essay-answer-like.ts:
```diff
@@ -0,0 +1,100 @@
+      },
+      isIndexed: 'unique',
+      db: {
+        isNullable: true,
```
question:
Since this is the unique indexed field, maybe it should be non-nullable?

nickhsine replied:
為了建立方便測試資料，目前先保留 Default NULL 的設定，之後等前端串接完畢，可以再考慮是不是要把 Default NULL 移除，確保資料乾淨。

ericsen-tsai reviewed (approved):
LGTM, thanks!

nickhsine committed:
refactor(cms): extract RoleEnum to constants and add session/type barrels; narrow memberId access

- Move RoleEnum to packages/cms/constants/role-enum.ts and re-export via lists/utils/access-control-list.
- Update imports/usages in keystone and ACL utilities to reference the new constant.
- Add CMS types modules and barrels (types/index.ts, keystone-session.ts, keystone-context.ts, keystone-type-info.ts) and adjust list types export.
- Fix TypeScript union error in lists/utils/member-owned-access.ts by narrowing session.data before reading memberId.
- Delete types/context.ts. Import { Context } from './types/index' instead.
- No runtime behavior changes; simplifies role imports and improves type safety.

nickhsine commented:
@ericsen-tsai  新的 commit [0a07a3a](https://github.com/kids-reporter/kids-reporter-monorepo/pull/788/commits/0a07a3a5463f9508907c6e92a7ac66a844ecc523) 再麻煩幫忙看看，謝謝

ericsen-tsai commented on packages/cms/types/keystone-context.ts:
```diff
@@ -0,0 +1,3 @@
+import type { Context } from '.keystone/types'
+
+export type { Context }
```
question: since there is session object inside the context, maybe you want to export it as `Context<Session>`?

nickhsine replied:
Address in [12a6e02](https://github.com/kids-reporter/kids-reporter-monorepo/pull/788/commits/12a6e02264f6a3c284febd97508acac38e4074d0)

nickhsine committed:
chore(cms): add migrations/20251028072549_add_choice_and_essay_answer_lists
